### PR TITLE
Improvements and bug fixes for the character selector screen

### LINKED ISSUE #6436: Issues with the new Character Creator / Selector
Here is a statement of accounts of my experience on the latest staging as of this issue:
- right after booting up the game, first character (out of two, in my case) has no preview.
- after clicking `Delete Character` for this first character:
    - there's no warning modal about deleting a character.
    - the count of how many characters I have is not updated.
    - I am given the `Please create a new character` pane instead of switching to the next character, if available.
    - after clicking the change character arrows, the character count is now updated, but I am stuck on the aforementioned screen. I can't select the other character.
    - the edit button is available on this  `Please create a new character` pane - if I click it, I am able to edit the second character.
        - if I hit exit, I am not returning back to the character select list, but instead launching into the server. This is disorientating. Swap the Exit and the Back button positions.
    - the character isn't actually deleted: hit Exit, close and reboot the game -> character returns.
- there's no obvious way to proceed: `Exit` isn't a good name (I might think it means to exit the game - after all, we never clicked on anything to get here right after booting the game). `Continue` would be better and needs a better location, too.
- after loading into the game, the character I spawn with is completely different to the "deleted" character and the remaining character.

### PR DESCRIPTION
**Pull request by MaxIsJoe**
### Purpose
Part of #6276 
Fixes a lot of hidden/rare bugs when loading and deleting characters, New look for the character selector, dropdown menu and other small improvements have been added.

image

### Notes:
closes #6436 as 99% of all the issues reported there has been fixed.

also tickes two boxes off in #6195 where it says character UI is not pixel perfect

### Changelog:

CL: [Improvement] - The character selector has had some UI changes.
CL: [Improvement] - The exit button is now a cancel button, returning you to the character selector when you don't want to save your changes.
Cl: [Improvement] - The character selector now has a pixel perfect character preview.
CL: [New] - You can now see and select your characters via a drop menu.
CL: [New] - Added confirmation dialogue for when you click on the Delete Character Button.
CL: [Fix] - Players will no longer be able to edit characters when the PlayerList is empty.
CL: [Fix] - The game will no longer try to spawn in ghost character sheets who have been deleted.
CL: [Fix] - You no longer can create one-time character sheets by pressing the Finish button on the top.
CL: [Fix] - You no longer have the ability to join the round if you have no characters. Previously people had this ability using deleted characters.
CL: [Fix] - Fixed a collection of bugs when loading/editing characters that caused a bunch of silent issues.
CL: [Fix] - If you're savefile is corrupted the game won't freeze you on the Pre-round screen and will now try correcting your savefile to continue playing.
CL: [Fix] - When editing a character the game will load in your current character index.
CL: [Fix] - Fixed a bug where creating a new character caused the game to not register the last character sheet correctly.

### CODE REVIEW

**Comment by NoooneyDude** [comment]
I had some issues playing a local build of this (couldn't enter the game), but this might be just local to me. Could someone else give testing this a go using a built game?

For the parts that I _could_ test:

- I can't ready-up in the editor:
lobby
no errors.
- In the built game (possibly editor too) I get some errors when clicking on the `Character` button from the login screen. Might be as simple as turning off networked sprites on `SpriteHandler`.
character button
errors


It is good to see a lot of the character selector issues have been resolved.

**Comment by MaxIsJoe** [comment]
> * I can't ready-up in the editor

This sounds like the PlayerCharacters list being empty because unlike currentCharacter the game doesn't load that list on awake.
image


> * In the built game (possibly editor too) I get some errors when clicking on the `Character` button from the login screen. Might be as simple as turning off networked sprites on `SpriteHandler`.

This existed waaay before this bounty and health update, it's a harmless error that does nothing except telling us that "hey, this sprite is not networked". It's not important and does nothing.

**Comment by NoooneyDude** [comment]
I had a character, so I shouldn't need to do anything extra.

**Comment by MaxIsJoe** [comment]
> I had a character, so the count shouldn't be empty.

Have you opened the character selector before readying up?

**Comment by NoooneyDude** [comment]
No, and you shouldn't need to.

**Comment by MaxIsJoe** [comment]
> No, and you shouldn't need to.

I was describing the bug, I'll fix it later when I have time but I'm letting you know what's the issue at the moment.

**Comment by NoooneyDude** [comment]
Thank you.

**Comment by MaxIsJoe** [comment]
Why is it giving out errors? It compiles just fine on my end.

**Comment by MaxIsJoe** [comment]
Soooooo, any news for why it's still giving a tests error? everything works fine on my end.

**Comment by ThatDan123** [comment]
Try merging upstream

**Comment by MaxIsJoe** [comment]
> Try merging upstream

merge or rebase? this is a very important thing because both have totally different consequences

**Comment by ThatDan123** [comment]
merge

**Comment by MaxIsJoe** [comment]
No errors, we can merge finally yay!

**Comment by NoooneyDude** [comment]
- Getting unpopulated preview:
unpopulated previeww
Associative NRE:
```
ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
System.ThrowHelper.ThrowArgumentOutOfRangeException (System.ExceptionArgument argument, System.ExceptionResource resource) (at <9577ac7a62ef43179789031239ba8798>:0)
System.ThrowHelper.ThrowArgumentOutOfRangeException () (at <9577ac7a62ef43179789031239ba8798>:0)
System.Collections.Generic.List`1[T].get_Item (System.Int32 index) (at <9577ac7a62ef43179789031239ba8798>:0)
UI.CharacterCreator.CharacterCustomization.RefreshSelectorData () (at Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs:269)
UI.CharacterCreator.CharacterCustomization.GetSavedCharacters () (at Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs:1066)
UI.CharacterCreator.CharacterCustomization.Awake () (at Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs:128)
UnityEngine.GameObject:SetActive(Boolean)
UI.GUI_PreRoundWindow:OnCharacterButton() (at Assets/Scripts/UI/Systems/PreRound/GUI_PreRoundWindow.cs:185)
```
- Another one at line 432 when editing the character.

- Clicking an arrow seems to load the character properly:
character loaded
But when I edit that character I get a completely different, randomly generated one:
editing new character
- If I spawn, I spawn as this random character and not the `Clarke Hanford` selected in the selection window.
- However, one time I couldn't spawn and the server rejected my spawn requests, so I was stuck on the job selection window. 
- After creating a new character, the the character name and the preview don't match:
bad preview

**Comment by MaxIsJoe** [comment]
> Getting unpopulated preview

> But when I edit that character I get a completely different, randomly generated one

> After creating a new character, the the character name and the preview don't match

I guess this is happening because of the new dropdown menus which uses a different way for handling lists? Idk, I haven't encountered this issue at all during testing as it has (or was) already been fixed in previous PRs. Gonna take a look at it tomorrow because right now I'm busy with highschool.

**Comment by MaxIsJoe** [comment]
I couldn't replicate a single issue you mentioned @NoooneyDude 
Everything seems to be working fine and even the code that I wrote should prevent this sort of issues.

**Comment by MaxIsJoe** [comment]
These two new commits I introduced are just two things that were on my todo list but I didn't have time to fix them.
They're very minor changes but are welcome non-the-less. I'm still going for another test run to see if I can somehow trigger any more issues that need to be fixed.

**Comment by MaxIsJoe** [comment]
Alright after a bunch of testing I found only one issue

For some reason the game keeps loading the character that's behind the one you have selected. I guess this have returned while adding the dropdown menu and I don't know how I didn't notice this until now. gonna try fixing this and see how this goes.

**Comment by MaxIsJoe** [comment]
Alright some very good news;

I fixed the issue where you spawn with the wrong character and I've discovered a veeeeeeerrrry nasty collection of bugs that nobody noticed including me because of how silent and un-noticable they are. The character creator should now function more in it's intended behavior when loading/creating characters and we should have less bugs now when loading and creating characters.

If anyone wants to test it one more time then feel free to go.

**Comment by PetMudstone** [comment]
I did some testing, message thread here: https://discord.com/channels/273774715741667329/312454684021620736/841546352922984488
I would summarize my findings if I wasn't currently horribly confused and also tired.

**Comment by MaxIsJoe** [comment]
Alright, one more test needed.

**Comment by NoooneyDude** [comment]
I've had enough of reporting the same bugs every time.

bugs

**Comment by MaxIsJoe** [comment]
> I've had enough of reporting the same bugs every time.
> 
> bugs

............ Are you even updating your local branch of this PR?
This is outdated. We're no longer using that character preview.

I have no issues on my end doing anything.

**Comment by NoooneyDude** [comment]
> ............ Are you even updating your local branch of this PR?
> This is outdated. We're no longer using that character preview.

Yes, this your latest version pulled right from your fork and built.

**Comment by MaxIsJoe** [comment]
> > ............ Are you even updating your local branch of this PR?
> > This is outdated. We're no longer using that character preview.
> 
> Yes, this your latest version pulled right from your fork and built.

No, No it is not.

**Comment by MaxIsJoe** [comment]
> > > ............ Are you even updating your local branch of this PR?
> > > This is outdated. We're no longer using that character preview.
> > 
> > 
> > Yes, this your latest version pulled right from your fork and built.
> 
> No, No it is not.

All image preview code has been removed and all of the issues that were reported has been fixed.
if this was TRULY the updated branch on your end, the game would straight up not work because the code to show the preview images have been removed completely.
Update your branch properly then come back again.

**Comment by NoooneyDude** [comment]
You're right. My sincere apologies.

**Comment by MaxIsJoe** [comment]
It is done.

This should be the last problem fixed completely. If you want to test it then go for it. other than that I think this is 100% ready to merge this time for reals (not clickbait)

File: UnityProject/Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs [hunk 1]
```diff
@@ -200,105 +213,107 @@ private void ShowCharacterCreator()
...
 		/// <summary>
-		/// Loads an image from a localpath or from a server.
+		/// Whenever the player changes his character via the dropdown menu we make sure that currentCharacterIndex is set accordingly
+		/// And then we make sure that the currentCharacter is also loaded in.
+		/// Note : to unify the way loading character data is; we mainly use ItemChange now for everything to make bug trackign less and code better.
 		/// </summary>
-		/// <param name="path">The path to the image that will be used as a texture.</param>
-		/// <returns></returns>
-		private IEnumerator<UnityWebRequestAsyncOperation> GetPreviewImage(string path)
+		/// <param name="newValue"></param>
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs [hunk 2]
```diff
@@ -200,105 +213,107 @@ private void ShowCharacterCreator()
...
-					CharacterPreviewImg.SetActive(false);
-					NoPreviewError.SetActive(true);
-				}
-				else
-				{
-					CharacterPreviewImg.texture = DownloadHandlerTexture.GetContent(uwr);
-				}
+		private void CheckIfCharacterListIsEmpty()
+		{
+			if(PlayerCharacters.Count == 0)
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs]:
```suggestion
			if (PlayerCharacters.Count == 0)
```

File: UnityProject/Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs [hunk 3]
```diff
@@ -200,105 +213,107 @@ private void ShowCharacterCreator()
...
+
+		public void HideCharacterDeletionConfirmation()
+		{
+			DeleteCharacterButton.SetActive(true);
+			ConfirmDeleteCharacterObject.SetActive(false);
+		}
+
 		public void CreateCharacter()
 		{
+			if(currentCharacter != null)
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs]:
```suggestion
			if (currentCharacter != null)
```

File: UnityProject/Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs [hunk 4]
```diff
@@ -968,36 +978,51 @@ private void SaveCharacters()
...
 			string path = Application.persistentDataPath + "characters.json";
+			if (PlayerCharacters.Count == 0)
+			{
+				json = "";
+			}
+			else
+			{
+				json = JsonConvert.SerializeObject(PlayerCharacters, settings);
+			}
 			if(File.Exists(path))
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/UI/Systems/Lobby/CharacterCustomization.cs]:
```suggestion
			if (File.Exists(path))
```

**Comment by MaxIsJoe** [comment]
oh, I was just going to push a fix for the character escapee thing.
well I guess there's no harm in moving it to a new PR.

**Pull request merged**